### PR TITLE
Fix issue in let expressions with lambda functions

### LINKED ISSUE #35015: Compiler crashes for let-expr combined with anonymous function
**Description:**
$subject

**Steps to reproduce:**
```ballerina
public function main() {
    var _ = let null n = null in function() returns boolean {
        return n == ();
    };
}
```

<details>
  <summary>Stack trace.(Click to expand)</summary>

```
[2022-02-10 17:13:18,183] SEVERE {b7a.log.crash} - Index 0 out of bounds for length 0 
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.createFunctionInfoArray(MainMethodGen.java:341)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.loadCLIArgsForMain(MainMethodGen.java:298)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.generateUserMainFunctionCall(MainMethodGen.java:273)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.generateMainMethod(MainMethodGen.java:167)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:428)
	at java.base/java.util.HashMap$EntrySet.forEach(HashMap.java:1039)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:397)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:808)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:95)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:70)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:301)
	at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:477)
	at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:390)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:203)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:51)
```
</details>

**Affected Versions:**
Ballerina 2201.0.0 (Swan Lake)

### LINKED ISSUE #35030: Compiler crashes in let-expr for query-expr
**Description:**
$subject

**Steps to reproduce:**
```
public function main() {
    var a = let int n = 2 in from var i in [1, 3] 
                                where i == n
                                select i;
    io:println(a);
}
```

This does not crash without the `where` clause.

<details>
  <summary>Stack trace.(Click to expand)</summary>

```
[2022-02-11 12:52:29,510] SEVERE {b7a.log.crash} - Index 0 out of bounds for length 0 
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.createFunctionInfoArray(MainMethodGen.java:341)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.loadCLIArgsForMain(MainMethodGen.java:298)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.generateUserMainFunctionCall(MainMethodGen.java:273)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MainMethodGen.generateMainMethod(MainMethodGen.java:167)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:428)
	at java.base/java.util.HashMap$EntrySet.forEach(HashMap.java:1039)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:397)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:808)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:95)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:70)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:301)
	at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:477)
	at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:390)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:203)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:51)
```
</details>

**Affected Versions:**
Ballerina 2201.0.0 (Swan Lake)

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #35030
Fixes #35015

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java [hunk 1]
```diff
@@ -1906,9 +1906,13 @@ public void visit(BLangAnnotAccessExpr annotAccessExpr) {
     public void visit(BLangStatementExpression bLangStatementExpression) {
         if (bLangStatementExpression.stmt.getKind() == NodeKind.BLOCK) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java:1907]: Shall we extract `bLangStatementExpression.stmt` out to a variable?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java [hunk 2]
```diff
@@ -1904,14 +1904,19 @@ public void visit(BLangAnnotAccessExpr annotAccessExpr) {
...
+        if (exprStmt.getKind() == NodeKind.BLOCK) {
+            BLangBlockStmt bLangBlockStmt = (BLangBlockStmt) exprStmt;
+            if (bLangBlockStmt.isLetExpr) {
+                bLangStatementExpression.stmt = rewrite(bLangBlockStmt, env);
+            } else {
+                for (int i = 0; i < bLangBlockStmt.stmts.size(); i++) {
+                    BLangStatement stmt = bLangBlockStmt.stmts.remove(i);
+                    bLangBlockStmt.stmts.add(i, rewrite(stmt, env));
+                }
             }
```
Comment (lochana-chathura) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java:1917]: HBT this? we can avoid the nested if-else.
```suggestion
        if (exprStmt.getKind() == NodeKind.BLOCK && !((BLangBlockStmt) exprStmt).isLetExpr) {
            BLangBlockStmt bLangBlockStmt = (BLangBlockStmt) exprStmt;
            for (int i = 0; i < bLangBlockStmt.stmts.size(); i++) {
                BLangStatement stmt = bLangBlockStmt.stmts.remove(i);
                bLangBlockStmt.stmts.add(i, rewrite(stmt, env));
            }
```

**Review by lochana-chathura — APPROVED** [review]
LGTM

**Comment by lasinicl** [comment]
There's a regression issue with the fix done here. Looking into it now. Closing temporarily. Will reopen once it is fixed.

**Comment by lasinicl** [comment]
Opened as a new PR to master https://github.com/ballerina-platform/ballerina-lang/pull/35796

**Pull request closed without merging**